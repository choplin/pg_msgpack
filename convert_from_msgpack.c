#include <inttypes.h>

#include "postgres.h"
#include "lib/stringinfo.h"

#include "convert_from_msgpack.h"

/*
 * private functions for msgpack_to_json_string
 */
static void msgpack_object_to_string(msgpack_object o, StringInfo out);
static inline void raw_to_string(msgpack_object o, StringInfo out);
static void array_to_string(msgpack_object o, StringInfo out);
static void map_to_string(msgpack_object o, StringInfo out);


char *
msgpack_to_json_string(msgpack_object o)
{
	StringInfoData out;
	initStringInfo(&out);

	msgpack_object_to_string(o, &out);

	return out.data;
}

bytea *
msgpack_to_bytea(msgpack_object o)
{
	bytea			*out;
	msgpack_sbuffer	*sbuf;
	msgpack_packer	*pk;

	sbuf = msgpack_sbuffer_new();

	pk = msgpack_packer_new(sbuf, msgpack_sbuffer_write);
	msgpack_pack_object(pk, o);

	out = (bytea *) palloc(sbuf->size + VARHDRSZ);
	SET_VARSIZE(out, sbuf->size + VARHDRSZ);
	memcpy(VARDATA(out), sbuf->data, sbuf->size);

	msgpack_packer_free(pk);
	msgpack_sbuffer_free(sbuf);

	return out;
}

/*
 * private functions
 */
static inline void
msgpack_object_to_string(msgpack_object o, StringInfo out)
{
	switch(o.type) {
	case MSGPACK_OBJECT_NIL:
		appendStringInfoString(out, "null");
		break;

	case MSGPACK_OBJECT_BOOLEAN:
		appendStringInfoString(out, (o.via.boolean ? "true" : "false"));
		break;

	case MSGPACK_OBJECT_POSITIVE_INTEGER:
		appendStringInfo(out, "%"PRIu64, o.via.u64);
		break;

	case MSGPACK_OBJECT_NEGATIVE_INTEGER:
		appendStringInfo(out, "%"PRIi64, o.via.i64);
		break;

	case MSGPACK_OBJECT_DOUBLE:
		appendStringInfo(out, "%f", o.via.dec);
		break;

	case MSGPACK_OBJECT_RAW:
		raw_to_string(o, out);
		break;

	case MSGPACK_OBJECT_ARRAY:
		array_to_string(o, out);
		break;

	case MSGPACK_OBJECT_MAP:
		map_to_string(o, out);
		break;

	default:
		appendStringInfo(out, "#<UNKNOWN %i %"PRIu64">", o.type, o.via.u64);
	}
}

static inline void
raw_to_string(msgpack_object o, StringInfo out)
{
	char *buf;
	
	buf = palloc(o.via.raw.size + 1);
	memcpy(buf, o.via.raw.ptr, o.via.raw.size);
	buf[o.via.raw.size] = '\0';

	appendStringInfoChar(out, '"');
	appendStringInfoString(out, buf);
	appendStringInfoChar(out, '"');

	pfree(buf);
}

static void
array_to_string(msgpack_object o, StringInfo out)
{
	msgpack_object *p;
	msgpack_object *pend;

	appendStringInfoChar(out, '[');

	if (o.via.array.size != 0) {
		p = o.via.array.ptr;
		pend =  p + o.via.array.size;

		msgpack_object_to_string(*p, out);

		++p;

		for(; p < pend; ++p) {
			appendStringInfoString(out, ", ");
			msgpack_object_to_string(*p, out);
		}
	}

	appendStringInfoChar(out, ']');
}

static void
map_to_string(msgpack_object o, StringInfo out)
{
	msgpack_object_kv *p;
	msgpack_object_kv *pend;

	appendStringInfoChar(out, '{');
	if (o.via.map.size != 0) {
		p = o.via.map.ptr;
		pend = p + o.via.map.size;

		msgpack_object_to_string(p->key, out);
		appendStringInfoChar(out, ':');
		msgpack_object_to_string(p->val, out);

		++p;

		for (; p < pend; ++p) {
			appendStringInfoString(out, ", ");

			msgpack_object_to_string(p->key, out);
			appendStringInfoChar(out, ':');
			msgpack_object_to_string(p->val, out);
		}
	}
	appendStringInfoChar(out, '}');
}
