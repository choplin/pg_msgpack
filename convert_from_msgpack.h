#ifndef __CONVERT_FROM_MSGPACK__
#define __CONVERT_FROM_MSGPACK__

#include <msgpack.h>

#include "postgres.h"

/* Convert msgpack_object to string */
char * msgpack_to_json_string(msgpack_object o);

/* Convert msgpack_object to bytea */
bytea * msgpack_to_bytea(msgpack_object o);

#endif /* __CONVERT_FROM_MSGPACK__ */
