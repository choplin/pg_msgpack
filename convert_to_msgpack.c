#include <errno.h>
#include <stdlib.h>
#include <string.h>

#include "postgres.h"
#include "utils/jsonapi.h"
#include "utils/builtins.h"

#include "convert_to_msgpack.h"

/*
 * Types for a scalar
 */
typedef enum {
	JSON_VALUE_SCALAR,
	MSGPACK_BINARY
} JsonValueType;

typedef struct {
	JsonTokenType	token_type;
	char			*token;
} JsonValueScalar;

typedef struct {
	size_t	size;
	char	*data;
} MsgpackBinary;

typedef union {
	JsonValueScalar	scalar;
	MsgpackBinary	msg;
} JsonValueUnion;

typedef struct {
	JsonValueType	type;
	JsonValueUnion	via;
} JsonValueData, *JsonValue;

/*
 * Types for a container
 */
typedef enum {
	JSON_ARRAY,
	JSON_OBJECT,
	JSON_TOP_LEVEL
} JsonContainerType;

typedef struct _JsonObjectField {
	struct _JsonObjectField	*next;
	char					*key;
	JsonValue				value;
} JsonObjectFieldData, *JsonObjectField;

typedef struct {
	size_t			length;
	JsonObjectField	start;
	JsonObjectField	end;
} JsonObject;

typedef struct _JsonArrayElement {
	struct _JsonArrayElement	*next;
	JsonValue			value;
} JsonArrayElementData, *JsonArrayElement;

typedef struct {
	size_t				length;
	JsonArrayElement	start;
	JsonArrayElement	end;
} JsonArray;

typedef struct {
	JsonValue	value;
} JsonTopLevel;

typedef union {
	JsonObject		object;
	JsonArray		array;
	JsonTopLevel	top_level;
} JsonContainerUnion;

typedef struct _JsonContainer{
	struct _JsonContainer		*parent;
	JsonContainerType			type;
	JsonContainerUnion			via;
} JsonContainerData, *JsonContainer;

/*
 * State for json_string_to_msgpack
 */
typedef struct {
	JsonLexContext	*lex;
	JsonContainer	current_container;
	msgpack_packer	*pk;
	msgpack_sbuffer	*buf;
} PackStateData, *PackState;

/*
 * Semantic action functions for json_string_to_msgpack
 */
static void sem_object_start(void *state);
static void sem_object_end(void *state);
static void sem_array_start(void *state);
static void sem_array_end(void *state);
static void sem_object_field_start(void *state, char *fname, bool isnull);
static void sem_array_element_start(void *state, bool isnull);
static void sem_scalar(void *state, char *token, JsonTokenType tokentype);

/*
 * Destroy function
 */

static inline void destroy_container(JsonContainer container);
static inline void destroy_object(JsonObject object);
static inline void destroy_array(JsonArray array);
static inline void destroy_value(JsonValue value);

/*
 * Pack functions
 */
static inline void pack_value(msgpack_packer *pk, const JsonValue value);
static inline void pack_scalar(msgpack_packer *pk, const JsonValueScalar scalar);
static inline void pack_string(msgpack_packer *pk, const char *str);
static inline void pack_number(msgpack_packer *pk, const char *number_token);
static inline void pack_unsigned_integer(msgpack_packer *pk, const char *unsigned_integer_token);
static inline void pack_integer(msgpack_packer *pk, const char *integer_token);
static inline void pack_real(msgpack_packer *pk, const char *real_token);

/*
 * Utility functions
 */
static inline JsonValue value_to_be_set(const JsonContainer container);

void
json_string_to_msgpack(const char *json_str, msgpack_sbuffer *buf)
{
	PackState		state;
	JsonLexContext	*lex = makeJsonLexContext(cstring_to_text(json_str), true);
	JsonSemAction	*sem;
	msgpack_packer	*pk;
	JsonContainer	top_level;
	JsonValue		top_level_value;

	/* initialize packer */
	pk = msgpack_packer_new(buf, msgpack_sbuffer_write);

	/* initialize top level object */
	top_level_value = palloc(sizeof(JsonValueData));

	top_level = palloc(sizeof(JsonContainerData));
	top_level->type = JSON_TOP_LEVEL;
	top_level->parent = NULL;
	top_level->via.top_level.value = top_level_value;

	/* initialize state object */
	state = palloc(sizeof(PackStateData));
	state->lex = lex;
	state->current_container = top_level;
	state->pk = pk;
	state->buf = buf;

	/* initialize sem action */
	sem = palloc(sizeof(JsonSemAction));
	sem->semstate 				= (void *) state;
	sem->object_start 			= sem_object_start;
	sem->object_end 			= sem_object_end;
	sem->array_start 			= sem_array_start;
	sem->array_end 				= sem_array_end;
	sem->object_field_start 	= sem_object_field_start;
	sem->object_field_end 		= NULL;
	sem->array_element_start 	= sem_array_element_start;
	sem->array_element_end 		= NULL;
	sem->scalar 				= sem_scalar;

	/* run parser */
	pg_parse_json(lex, sem);

	/* pack top level value */
	pack_value(pk, top_level_value);

	/* destroy packer */
	msgpack_packer_free(pk);

	/* destroy top level object */
	destroy_container(top_level);
}

static void
sem_object_start(void *state)
{
	PackState		_state = (PackState) state;
	JsonContainer	current_container, new_container;

	current_container = _state->current_container;

	/* create and set new container */
	new_container = palloc(sizeof(JsonContainerData));
	new_container->type = JSON_OBJECT;
	new_container->parent = current_container;
	new_container->via.object.length = 0;
	new_container->via.object.start = NULL;
	new_container->via.object.end = NULL;

	/* step forward to new container */
	_state->current_container = new_container;
}

static void
sem_object_end(void *state)
{
	PackState		_state = (PackState) state;
	msgpack_sbuffer	*buf;
	JsonContainer	current_container, parent_container;
	msgpack_packer 	*pk;
	JsonObjectField	field;
	size_t			packed_size;
	char			*packed_data;
	JsonValue		value;

	current_container = _state->current_container;
	pk = _state->pk;

	/* start to pack map */
	msgpack_pack_map(pk, current_container->via.object.length);

	/* pack each field */
	field = current_container->via.object.start;
	while (field != NULL) {
		/* pack key */
		pack_string(pk, field->key);
		/* pack value */
		pack_value(pk, field->value);
		/* next field*/
		field = field->next;
	}

	/* retrieve packed binary and reset buffer of packer */
	buf = _state->buf;
	packed_size = buf->size;
	packed_data = msgpack_sbuffer_release(buf);

	/* set packed binary to the last value of parent conationer */
	parent_container = current_container->parent;
	value = value_to_be_set(parent_container);
	value->type = MSGPACK_BINARY;
	value->via.msg.size = packed_size;
	value->via.msg.data = packed_data;

	/* step back to parent level */
	_state->current_container = parent_container;
	destroy_container(current_container);
}

static void
sem_array_start(void *state)
{
	PackState		_state = (PackState) state;
	JsonContainer	current_container, new_container;

	current_container = _state->current_container;

	/* create and set new container */
	new_container = palloc(sizeof(JsonContainerData));
	new_container->type = JSON_ARRAY;
	new_container->parent = current_container;
	new_container->via.array.length = 0;
	new_container->via.array.start = NULL;
	new_container->via.array.end = NULL;

	/* step forward to new container */
	_state->current_container = new_container;
}

static void
sem_array_end(void *state)
{
	PackState			_state = (PackState) state;
	msgpack_sbuffer		*buf;
	JsonContainer		current_container, parent_container;
	msgpack_packer 		*pk;
	JsonArrayElement	element;
	size_t				packed_size;
	char				*packed_data;
	JsonValue			value;

	current_container = _state->current_container;
	pk = _state->pk;

	/* start to pack array */
	msgpack_pack_array(pk, current_container->via.array.length);

	/* pack each element */
	element = current_container->via.array.start;
	while (element != NULL) {
		/* pack value */
		pack_value(pk, element->value);
		/* next element */
		element = element->next;
	}

	/* retrieve packed binary and reset buffer of packer */
	buf = _state->buf;
	packed_size = buf->size;
	packed_data = msgpack_sbuffer_release(buf);

	/* set packed binary to the last value of parent conationer */
	parent_container = current_container->parent;
	value = value_to_be_set(parent_container);
	value->type = MSGPACK_BINARY;
	value->via.msg.size = packed_size;
	value->via.msg.data = packed_data;

	/* step back to parent level */
	_state->current_container = parent_container;
	destroy_container(current_container);
}

static void
sem_object_field_start(void *state, char *fname, bool isnull)
{
	PackState		_state = (PackState) state;
	JsonContainer	container;
	JsonObjectField	new_field;
	JsonValue		new_value;

	container = _state->current_container;

	new_value = palloc(sizeof(JsonValueData));

	new_field = palloc(sizeof(JsonObjectFieldData));
	new_field->next = NULL;
	new_field->key = pstrdup(fname);
	new_field->value = new_value;

	container->via.object.length += 1;

	if (container->via.object.end == NULL) {
		container->via.object.start = new_field;
		container->via.object.end = new_field;
	} else {
		container->via.object.end->next = new_field;
		container->via.object.end = new_field;
	}
}

static void
sem_array_element_start(void *state, bool isnull)
{
	PackState			_state = (PackState) state;
	JsonContainer		container;
	JsonArrayElement	new_element;
	JsonValue			new_value;

	container = _state->current_container;

	new_value = palloc(sizeof(JsonValueData));

	new_element = palloc(sizeof(JsonArrayElementData));
	new_element->next = NULL;
	new_element->value = new_value;

	container->via.array.length += 1;

	if (container->via.array.end == NULL) {
		container->via.array.start = new_element;
		container->via.array.end = new_element;
	} else {
		container->via.array.end->next = new_element;
		container->via.array.end = new_element;
	}
}

static void
sem_scalar(void *state, char *token, JsonTokenType tokentype)
{
	PackState	_state = (PackState) state;
	JsonContainer		container;
	JsonValue			value;

	container = _state->current_container;
	value = value_to_be_set(container);
	value->type = JSON_VALUE_SCALAR;
	value->via.scalar.token_type = tokentype;
	value->via.scalar.token = token;
}


static inline void
destroy_container(JsonContainer container)
{
	switch (container->type) {
		case JSON_OBJECT:
			destroy_object(container->via.object);
			break;
		case JSON_ARRAY:
			destroy_array(container->via.array);
			break;
		case JSON_TOP_LEVEL:
			destroy_value(container->via.top_level.value);
			break;
	}
	pfree(container);
}

static inline void
destroy_object(JsonObject object)
{
	JsonObjectField	field;

	field = object.start;
	while (field != NULL) {
		pfree(field->key);
		destroy_value(field->value);
		field = field->next;
	}
}

static inline void
destroy_array(JsonArray array)
{
	JsonArrayElement	element;

	element = array.start;
	while (element != NULL) {
		destroy_value(element->value);
		element = element->next;
	}
}

static inline void
destroy_value(JsonValue value)
{
	switch (value->type) {
		case JSON_VALUE_SCALAR:
			pfree(value->via.scalar.token);
			break;
		case MSGPACK_BINARY:
			free(value->via.msg.data);
			break;
	}
	pfree(value);
}

static inline void
pack_value(msgpack_packer *pk, const JsonValue value)
{
	switch (value->type) {
		case JSON_VALUE_SCALAR:
			pack_scalar(pk, value->via.scalar);
			break;
		case MSGPACK_BINARY:
			(* pk->callback)(pk->data, value->via.msg.data, value->via.msg.size);
			break;
	}
}

static inline void
pack_scalar(msgpack_packer *pk, const JsonValueScalar scalar)
{
	switch (scalar.token_type) {
		case JSON_TOKEN_STRING:
			pack_string(pk, scalar.token);
			break;
		case JSON_TOKEN_NUMBER:
			pack_number(pk, scalar.token);
			break;
		case JSON_TOKEN_TRUE:
			msgpack_pack_true(pk);
			break;
		case JSON_TOKEN_FALSE:
			msgpack_pack_false(pk);
			break;
		case JSON_TOKEN_NULL:
			msgpack_pack_nil(pk);
			break;
		default:
			/* TODO: parse error */
			break;
	}
}

static inline void
pack_string(msgpack_packer *pk, const char *str)
{
	size_t len = strlen(str);
	msgpack_pack_raw(pk, len);
	msgpack_pack_raw_body(pk, str, len);
}

static inline void
pack_number(msgpack_packer *pk, const char *number_token)
{
	if (strchr(number_token, '.') == NULL) {
		if (number_token[0] == '-')
			pack_integer(pk, number_token);
		else
			pack_unsigned_integer(pk, number_token);
	} else
		pack_real(pk, number_token);
}

static inline void
pack_unsigned_integer(msgpack_packer *pk, const char *uint_token)
{
	char *end;
	unsigned long value;
	
	errno = 0;
	value = strtoul(uint_token, &end, 10);
	/* TODO: error handling */

	/* TODO */
	msgpack_pack_unsigned_long(pk, value);
}

static inline void
pack_integer(msgpack_packer *pk, const char *int_token)
{
	char *end;
	long value;
	
	errno = 0;
	value = strtol(int_token, &end, 10);
	/* TODO: error handling */

	/* TODO */
	msgpack_pack_long(pk, value);
}

static inline void
pack_real(msgpack_packer *pk, const char *real_token)
{
	char *end;
	double value;

	errno = 0;
	value = strtod(real_token, &end);
	/* TODO: error handling */

	/* TODO */
	msgpack_pack_double(pk, value);
}

static inline JsonValue
value_to_be_set(const JsonContainer container)
{
	JsonValue value;
	switch (container->type) {
		case JSON_OBJECT:
			value = container->via.object.end->value;
			break;
		case JSON_ARRAY:
			value = container->via.array.end->value;
			break;
		case JSON_TOP_LEVEL:
			value = container->via.top_level.value;
			break;
	}
	return value;
};
