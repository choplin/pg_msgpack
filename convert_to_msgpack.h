#ifndef __CONVERT_TO_MSGPACK__
#define ___CONVERT_TO_MSGPACK__

/* TODO */
/* stop to load stdbool.h. force to use postgresql's bool */
#define _STDBOOL_H 
#define __STDBOOL_H 
#include <msgpack.h>

void json_string_to_msgpack(const char *json_str, msgpack_sbuffer *buf);

#endif /* ___CONVERT_TO_MSGPACK__ */
