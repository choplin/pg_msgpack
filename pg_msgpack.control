# msgpack extension
comment = 'msgpack'
default_version = '0.0.1'
module_pathname = '$libdir/pg_msgpack'
