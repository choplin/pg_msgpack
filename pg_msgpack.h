#ifndef __PG_MSGPACK_H__
#define __PG_MSGPACK_H__

#include "fmgr.h"

Datum msgpack_in(PG_FUNCTION_ARGS);
Datum msgpack_out(PG_FUNCTION_ARGS);
Datum msgpack_recv(PG_FUNCTION_ARGS);
Datum msgpack_send(PG_FUNCTION_ARGS);

#endif /* __PG_MSGPACK_H__ */
