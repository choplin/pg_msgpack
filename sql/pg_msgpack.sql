CREATE EXTENSION pg_msgpack;

-- conversion
SELECT '[null, true, false, 10, 5.500000, {"a":"b"}]'::msgpack;

-- operator
SELECT '{"a":"b"}'::msgpack -> 'a';
SELECT '[1,2,3]'::msgpack -> 1;
