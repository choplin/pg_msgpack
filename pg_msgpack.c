#include "postgres.h"
#include "libpq/pqformat.h"
#include "utils/bytea.h"

#include "pg_msgpack.h"
#include "convert_from_msgpack.h"
#include "convert_to_msgpack.h"

PG_MODULE_MAGIC;

PG_FUNCTION_INFO_V1(msgpack_in);
PG_FUNCTION_INFO_V1(msgpack_out);
PG_FUNCTION_INFO_V1(msgpack_recv);
PG_FUNCTION_INFO_V1(msgpack_send);

Datum
msgpack_in(PG_FUNCTION_ARGS)
{
	char	   		*json = PG_GETARG_CSTRING(0);
	bytea			*result;
	msgpack_sbuffer	*buf;

	if (json[0] == '\0')
		PG_RETURN_NULL();

	if (json[0] == '\\') {
		PG_RETURN_DATUM(DirectFunctionCall1(byteain, CStringGetDatum(json)));
	} else {
		buf = msgpack_sbuffer_new();
		json_string_to_msgpack(json, buf);

		result = (bytea *) palloc(buf->size + VARHDRSZ);
		SET_VARSIZE(result, buf->size + VARHDRSZ);
		memcpy(VARDATA(result), buf->data, buf->size);

		msgpack_sbuffer_free(buf);

		/* Internal representation is the same as bytea */
		PG_RETURN_BYTEA_P(result);
	}
}

Datum
msgpack_out(PG_FUNCTION_ARGS)
{
	bytea 				*data = PG_GETARG_BYTEA_P(0);
	char  				*out;
	msgpack_unpacked	msg;
	bool 				success;

	msgpack_unpacked_init(&msg);
	success = msgpack_unpack_next(&msg, VARDATA(data), VARSIZE(data) - VARHDRSZ, NULL);

	if (!success)
		PG_RETURN_NULL();

	out = msgpack_to_json_string(msg.data);

	msgpack_unpacked_destroy(&msg);

	PG_RETURN_CSTRING(out);
}

Datum
msgpack_recv(PG_FUNCTION_ARGS)
{
	StringInfo	buf = (StringInfo) PG_GETARG_POINTER(0);
	bytea		*result;
	int			nbytes;

	nbytes = buf->len - buf->cursor;
	result = (bytea *) palloc(nbytes + VARHDRSZ);
	SET_VARSIZE(result, nbytes + VARHDRSZ);

	pq_copymsgbytes(buf, VARDATA(result), nbytes);

	PG_RETURN_BYTEA_P(result);
}

Datum
msgpack_send(PG_FUNCTION_ARGS)
{
	bytea	*data = PG_GETARG_BYTEA_P_COPY(0);

	PG_RETURN_BYTEA_P(data);
}
