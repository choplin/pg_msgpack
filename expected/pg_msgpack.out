CREATE EXTENSION pg_msgpack;
-- conversion
SELECT '[null, true, false, 10, 5.500000, {"a":"b"}]'::msgpack;
                   msgpack                    
----------------------------------------------
 [null, true, false, 10, 5.500000, {"a":"b"}]
(1 row)

-- operator
SELECT '{"a":"b"}'::msgpack -> 'a';
 ?column? 
----------
 "b"
(1 row)

SELECT '[1,2,3]'::msgpack -> 1;
 ?column? 
----------
 2
(1 row)

